### DISCUSSION TITLE
How to get rgb depth from crest

### DISCUSSION THREAD

**Discussion by mooooh** [crest-ocean/crest#1071] [General]
Hello, I'm a student studying computer vision.

I found out about your crest github while looking for various tools to create a sea wave image dataset.

I would like to create a ground truth for the dataset by extracting depth for the sea wave image using your unity asset, Crest.

In Crest, I found AnimWaves expressing the wave height for the xyz value in the rgb channel.
![그림1](https://user-images.githubusercontent.com/42243173/154090131-10466efb-0bb4-4150-9476-f355492795e1.png)

Using this, I looked at the code LodDataMgrAnimWaves.cs because I wanted to see the height of the wave on the screen that the camera was looking at through the rgb channel, but I couldn't find the location of the camera when drawing AnimWaves.

AnimWaves seems to be an expression of the shape of the wave when viewed vertically from above, but I want to know the rgb channel of the height of the wave when viewed from the side to the horizontal plane.

As shown in the picture below, I would like to extract the height of the wave through the rgb channel.
![그림2](https://user-images.githubusercontent.com/42243173/154090201-e186c0a4-66e7-470f-bf16-8209f5017a52.png)

If you know how to do it or if you can help me, I look forward to hearing from you

**Comment by magenta404** [comment]
In master the following has been merged: #965 which might help. It will allow you to render the RGB values (Animated Waves option) onto the ocean surface. You could run a second camera with that enabled and target a render texture? That option samples the data in the vertex function so it won't be 100% accurate but should be close.

You'll need to remove the ` + 0.5` from the following line to make the colours full strength:
https://github.com/wave-harmonic/crest/blob/f2a1a18edfb28dcf3e1a5296aa1d17d0027d9d47/crest/Assets/Crest/Crest/Shaders/Ocean.shader#L735

**Comment by mooooh** [comment]
thank you so much!

**Comment by magenta404** [comment]
@mooh I made a mistake. The 0.5 is needed. The other issue is that the data texture stores real units and the output of the camera will be clamped between 0-1. Replacing that line with the following should work:

```shader
col.rgb = clamp(input.debugtint / 50, -1, 1) * 0.5 + 0.5;
```

The 50 is just the maximum distanced displacement value (in metres) so change as needed. And the 0.5 at the end ensures negative values are not clamped.

When you sample you'll need to convert it back to get real units.

Apologies. Hopefully got it right this time :^)

**Comment by mooooh** [comment]
I now understand it's RGB channels are the XYZ display on Animated Waves.
But I want to know height of the wave in which color. 
For example, if RGB value in a pixel is (244,253,4). Is this mean the G value (253) is the height of that pixel?

**Comment by magenta404** [comment]
> For example, if RGB value in a pixel is (244,253,4). Is this mean the G value (253) is the height of that pixel?

I explain this above. There's a couple ways you can get the height. What I described above is one way.

Another way is to sample the __CrestOceanMaskDepthTexture_ and reconstructed the world position from that (you'll have to work that out by googling). It is only available if the _Underwater Renderer_ is active (enable on that component: _Debug > Disable Height Above Water Optimization_).
